### PR TITLE
New mutable state task refresher for mutable state task rebuild

### PR DESCRIPTION
**Pull request by wxing1292**
* New mutable state task refresher which can re-generate tasks according to the current state of mutable state
* Add batch event ID to request cancel info & signal info for mutable state task refresher
* Format state builder
* Modifying mutableStateTaskGenerator accordingly
* Rename GetInFlightDecisionTask -> GetInFlightDecision
* Rename GetPendingDecision -> GetDecisionInfo
* Rename GetAllRequestCancels -> GetPendingRequestCancelInfos
* Rename GetAllSignalsToSend -> GetPendingSignalInfos
* Rename HasPendingDecisionTask -> HasPendingDecision
* Rename HasProcessedOrPendingDecisionTask -> HasProcessedOrPendingDecision

### CODE REVIEW

File: service/history/mutableState.go [hunk 1]
```diff
@@ -120,37 +120,38 @@ type (
 		GetChildExecutionInfo(int64) (*persistence.ChildExecutionInfo, bool)
 		GetChildExecutionInitiatedEvent(int64) (*workflow.HistoryEvent, bool)
 		GetCompletionEvent() (*workflow.HistoryEvent, bool)
+		GetDecisionInfo(int64) (*decisionInfo, bool)
```
Comment (longquanzheng) [inline service/history/mutableState.go]: The other renaming looks good except for this one. GetDecisionInfo(initiatedID) sounds like it is able to get any decision but it is actually getting current decision. I would prefer GetPendingDecisionInfo or GetCurrentDecisionInfo

Comment (longquanzheng) [reply service/history/mutableState.go]: There is already a GetPendingDecision(). What this does is Get and check if the decisionID is the same.

File: idl/github.com/uber/cadence/sqlblobs.thrift [hunk 2]
```diff
@@ -181,6 +181,7 @@ struct ChildExecutionInfo {
 
 struct SignalInfo {
   10: optional i64 (js.type = "Long") version
+  11: optional i64 (js.type = "Long") initiatedEventBatchID
```
Comment (longquanzheng) [inline idl/github.com/uber/cadence/sqlblobs.thrift]: Thanks for adding these

File: service/history/mutableState.go [hunk 3]
```diff
@@ -120,37 +120,38 @@ type (
...
 		GetLastFirstEventID() int64
 		GetLastWriteVersion() int64
 		GetNextEventID() int64
 		GetPreviousStartedEventID() int64
-		GetPendingDecision(int64) (*decisionInfo, bool)
 		GetPendingActivityInfos() map[int64]*persistence.ActivityInfo
 		GetPendingTimerInfos() map[string]*persistence.TimerInfo
 		GetPendingChildExecutionInfos() map[int64]*persistence.ChildExecutionInfo
+		GetPendingRequestCancelInfos() map[int64]*persistence.RequestCancelInfo
+		GetPendingSignalInfos() map[int64]*persistence.SignalInfo
```
Comment (longquanzheng) [inline service/history/mutableState.go]: Can we call it GetPendingSignalExternalInfos()? GetPendingSignalInfos sounds like getting received signals.

File: service/history/mutableStateBuilder.go [hunk 4]
```diff
@@ -4258,7 +4282,7 @@ func (e *mutableStateBuilder) closeTransactionHandleDecisionFailover(
 	}
 
 	// Handling mutable state turn from standby to active, while having a decision on the fly
-	di, ok := e.GetInFlightDecisionTask()
+	di, ok := e.GetInFlightDecision()
```
Comment (longquanzheng) [inline service/history/mutableStateBuilder.go]: nit: di=> decisionInfo

File: service/history/mutableStateTaskGenerator.go [hunk 5]
```diff
@@ -112,14 +117,11 @@ func newMutableStateTaskGenerator(
...
+	nowTimestamp int64,
 	event *shared.HistoryEvent,
 ) error {
 
-	// TODO when refactoring workflow reset functionality,
-	//  the logic below should be changed slightly, i.e. 'now' should not be derived from
-	//  start event, but the reset time
-
-	now := time.Unix(0, event.GetTimestamp())
+	now := time.Unix(0, nowTimestamp)
```
Comment (longquanzheng) [inline service/history/mutableStateTaskGenerator.go]: I remember we want to use timeSource for testiability, is it possible to do it here?

File: service/history/mutableStateTaskRefresher.go [hunk 6]
```diff
@@ -0,0 +1,394 @@
...
+) error {
+
+	if !r.mutableState.HasPendingDecision() {
+		// no decision task at all
+		return nil
+	}
+
+	decision, ok := r.mutableState.GetPendingDecision()
+	if !ok {
+		return &shared.InternalServiceError{Message: "impossible case: cannot get pending decision"}
```
Comment (longquanzheng) [inline service/history/mutableStateTaskRefresher.go]: it's better to remove "impossible case" because it doesn't give any valuable information. As an error, then it returns here, it is already possible.

File: service/history/mutableStateTaskRefresher.go [hunk 7]
```diff
@@ -0,0 +1,394 @@
...
+			scheduleEvent,
+		); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *mutableStateTaskRefresherImpl) refreshTasksForCancelExternalWorkflow(
```
Comment (longquanzheng) [inline service/history/mutableStateTaskRefresher.go]: nit: use RequestCancel insteand of Cancel to be consistent.

File: service/history/mutableState.go (hunk 1 above)
Comment (wxing1292) [reply service/history/mutableState.go]: we actually do `GetActivityInfo` which returns activity info and a bool, just following the same pattern here

GetPendingDecisionInfo() basically just get the pending decision, like GetPendingActivityInfos

File: service/history/mutableState.go (hunk 3 above)
Comment (wxing1292) [reply service/history/mutableState.go]: sure

File: service/history/mutableStateBuilder.go (hunk 4 above)
Comment (wxing1292) [reply service/history/mutableStateBuilder.go]: sure

File: service/history/mutableStateTaskRefresher.go (hunk 6 above)
Comment (wxing1292) [reply service/history/mutableStateTaskRefresher.go]: what if we actually have a bug? handling impossible case (i.e. return error rather than silently ignore the error) helps debugging, and in the future, refactoring.

File: service/history/mutableStateBuilder.go (hunk 4 above)
Comment (wxing1292) [reply service/history/mutableStateBuilder.go]: changing di -> decision or otherwise will collide with the decisionInfo type

File: service/history/mutableStateTaskRefresher.go (hunk 7 above)
Comment (wxing1292) [reply service/history/mutableStateTaskRefresher.go]: sure

**Comment by wxing1292** [comment]
for other var renaming, e.g. ai -> acrtivityInfo, si -> signalInfo, rci -> requestCancelInfo will do another PR

File: service/history/mutableStateTaskRefresher.go (hunk 6 above)
Comment (longquanzheng) [reply service/history/mutableStateTaskRefresher.go]: I understand what you are trying to give. This error is more critical than other errors like ConditionFailed/Timeout/ShardLost etc. When this error happens, it means we could have a bug. 

Imagine you run a command line and it returns an error telling you this is impossible... Whether it is impossible or not is too subjective to tell at that point. What you need is provide the fact what is causing error.

So I would suggest to reword it to "cannot get pending decision, it could be a bug". It will be more objective.

Comment (wxing1292) [reply service/history/mutableStateTaskRefresher.go]: done
